File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/MetalOwnedTextureFrame.java
package org.maplibre.nativeffi.render;

/** Borrowed Metal texture frame valid only while its frame handle is open. */
public final class MetalOwnedTextureFrame {
  private final FrameScope scope;
  private final long generation;
  private final int width;
  private final int height;
  private final double scaleFactor;
  private final long frameId;
  private final NativePointer texture;
  private final NativePointer device;
  private final long pixelFormat;

  MetalOwnedTextureFrame(
      FrameScope scope,
      long generation,
      int width,
      int height,
      double scaleFactor,
      long frameId,
      NativePointer texture,
      NativePointer device,
      long pixelFormat) {
    this.scope = scope;
    this.generation = generation;
    this.width = width;
    this.height = height;
    this.scaleFactor = scaleFactor;
    this.frameId = frameId;
    this.texture = texture;
    this.device = device;
    this.pixelFormat = pixelFormat;
  }

  public long generation() {
    scope.ensureActive();
    return generation;
  }

  public int width() {
    scope.ensureActive();
    return width;
  }

  public int height() {
    scope.ensureActive();
    return height;
  }

  public double scaleFactor() {
    scope.ensureActive();
    return scaleFactor;
  }

  public long frameId() {
    scope.ensureActive();
    return frameId;
  }

  public NativePointer texture() {
    scope.ensureActive();
    return texture;
  }

  public NativePointer device() {
    scope.ensureActive();
    return device;
  }

  public long pixelFormat() {
    scope.ensureActive();
    return pixelFormat;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/render/MetalOwnedTextureFrame.kt
package org.maplibre.nativeffi.render

/** Borrowed Metal texture frame valid only while its frame handle is open. */
public class MetalOwnedTextureFrame
internal constructor(
  private val scope: FrameScope,
  private val generationValue: Long,
  private val widthValue: Int,
  private val heightValue: Int,
  private val scaleFactorValue: Double,
  private val frameIdValue: Long,
  private val textureValue: NativePointer,
  private val deviceValue: NativePointer,
  private val pixelFormatValue: Long,
) {
  /** Session generation preserved as a native `uint64_t` bit pattern in [Long]. */
  public fun generation(): Long = checked { generationValue }

  public fun width(): Int = checked { widthValue }

  public fun height(): Int = checked { heightValue }

  public fun scaleFactor(): Double = checked { scaleFactorValue }

  /** Opaque frame identity preserved as a native `uint64_t` bit pattern in [Long]. */
  public fun frameId(): Long = checked { frameIdValue }

  public fun texture(): NativePointer = checked { textureValue }

  public fun device(): NativePointer = checked { deviceValue }

  /** Backend-native Metal pixel format preserved as a native uint64 bit pattern. */
  public fun pixelFormat(): Long = checked { pixelFormatValue }

  private inline fun <T> checked(block: () -> T): T {
    scope.ensureActive()
    return block()
  }
}
